src/class/bth/bth_device.c

```diff
@@ -0,0 +1,238 @@
+
+bool tud_bt_event_send(void *event, uint16_t event_len)
+{
+  return bt_tx_data(_btd_itf.ep_ev, event, event_len);
+}
+
+bool tud_bt_acl_data_send(void *event, uint16_t event_len)
+{
+  return bt_tx_data(_btd_itf.ep_acl_in, event, event_len);
+}
+
+//--------------------------------------------------------------------+
+// USBD Driver API
+//--------------------------------------------------------------------+
+void btd_init(void)
+{
+  tu_memclr(&_btd_itf, sizeof(_btd_itf));
+}
+
+void btd_reset(uint8_t rhport)
+{
+  (void)rhport;
+}
+
+bool btd_open(uint8_t rhport, tusb_desc_interface_t const *itf_desc, uint16_t *p_length)
+{
+  tusb_desc_endpoint_t const *desc_ep;
+  // Ensure this is BT Primary Controller
+  TU_VERIFY(TUSB_CLASS_WIRELESS_CONTROLLER == itf_desc->bInterfaceClass &&
+            TUD_BT_APP_SUBCLASS == itf_desc->bInterfaceSubClass &&
+            TUD_BT_PROTOCOL_PRIMARY_CONTROLLER == itf_desc->bInterfaceProtocol);
+
+  // Distinguish interface by number of endpoints, as both interface have same class, subclass and protocol
+  if (itf_desc->bNumEndpoints == 3)
+  {
+    _btd_itf.itf_num = itf_desc->bInterfaceNumber;
+
+    desc_ep = (tusb_desc_endpoint_t const *) tu_desc_next(itf_desc);
+
+    TU_ASSERT(TUSB_DESC_ENDPOINT == desc_ep->bDescriptorType && TUSB_XFER_INTERRUPT == desc_ep->bmAttributes.xfer);
+    TU_ASSERT(dcd_edpt_open(rhport, desc_ep));
+    _btd_itf.ep_ev = desc_ep->bEndpointAddress;
+
+    // Open endpoint pair
+    TU_ASSERT(usbd_open_edpt_pair(rhport, tu_desc_next(desc_ep), 2, TUSB_XFER_BULK, &_btd_itf.ep_acl_out,
+                                  &_btd_itf.ep_acl_in));
+
+    // Prepare for incoming data from host
+    TU_ASSERT(usbd_edpt_xfer(rhport, _btd_itf.ep_acl_out, _btd_itf.epout_buf, CFG_TUD_BTH_DATA_EPSIZE), false);
+
+    (*p_length) += sizeof(tusb_desc_interface_t) + 3 * sizeof(tusb_desc_endpoint_t);
+  }
+  else if (itf_desc->bNumEndpoints == 2)
+  {
+    uint8_t dir;
+
+    desc_ep = (tusb_desc_endpoint_t const *) tu_desc_next(itf_desc);
+    TU_ASSERT(desc_ep->bDescriptorType == TUSB_DESC_ENDPOINT);
+    dir = tu_edpt_dir(desc_ep->bEndpointAddress);
+    _btd_itf.ep_voice[dir] = desc_ep->bEndpointAddress;
+    // Store endpoint size for alternative 0
+    _btd_itf.ep_voice_size[dir][0] = (uint8_t) desc_ep->wMaxPacketSize.size;
+
+    desc_ep = (tusb_desc_endpoint_t const *) tu_desc_next(desc_ep);
+    TU_ASSERT(desc_ep->bDescriptorType == TUSB_DESC_ENDPOINT);
+    dir = tu_edpt_dir(desc_ep->bEndpointAddress);
+    _btd_itf.ep_voice[dir] = desc_ep->bEndpointAddress;
+    // Store endpoint size for alternative 0
+    _btd_itf.ep_voice_size[dir][0] = (uint8_t) desc_ep->wMaxPacketSize.size;
+
+    for (int i = 1; i < CFG_TUD_BTH_ISO_ALT_COUNT; ++i) {
+      // Make sure rest of alternatives matches
+      itf_desc = (tusb_desc_interface_t const *)tu_desc_next(desc_ep);
+      TU_ASSERT(itf_desc->bDescriptorType == TUSB_DESC_INTERFACE);
```

Comment:
@kasjer I the ATL interface will be back to back, we could break whenever the descriptor type is not a Interface type, or alternate number is zero (next driver interface).